test/run_e2e_test.py

```diff
@@ -57,37 +61,58 @@ class SystemMetricsEnd2EndTestException(Exception):
 class StatisticsListener(Node):
     """Listen for MetricsMessages published on the /system_metrics topic."""
 
-    def __init__(self, future: Future, expected_lifecycle_nodes: List):
+    def __init__(self, future: Future, expected_lifecycle_nodes: List,
+                 expected_number_of_messages_to_receive: int):
         super().__init__('statisticsListener')
 
         self.sub = self.create_subscription(MetricsMessage,
                                             EXPECTED_TOPIC,
                                             self.listener_callback,
                                             QOS_DEPTH)
         self.future = future
-        self.received_all_published_stats = False
-        self.expected_lifecycle_nodes = expected_lifecycle_nodes
+        # key is node name, value is expected number of messages to receive
+        self.expected_lifecycle_nodes_dict = {}
+        for node in expected_lifecycle_nodes:
+            self.expected_lifecycle_nodes_dict[node] = expected_number_of_messages_to_receive
+        self.lock = Lock()
 
     def listener_callback(self, msg) -> None:
         """
         Handle published MetricsMessages.
 
-        Checks each received message measurement_source_name against a list of
-        expected_lifecycle_nodes and marks the future field as done if all sources
-        have been observed.
+        Checks each received message measurement_source_name against a dict of
+        expected_lifecycle_nodes, decrements each entry for its message received,
+        and marks the future field as done if all sources have been observed.
         :param msg: received message
         :return: None
         """
         node_name = '/' + msg.measurement_source_name
 
-        if node_name in self.expected_lifecycle_nodes:
+        if node_name in self.expected_lifecycle_nodes_dict:
             logging.debug('received message from %s', node_name)
-            self.expected_lifecycle_nodes.remove(node_name)
-
-        if not self.expected_lifecycle_nodes:
+            removed = False
+            with self.lock:
+                self.expected_lifecycle_nodes_dict[node_name] = (
+                        self.expected_lifecycle_nodes_dict[node_name] - 1)
+                if self.expected_lifecycle_nodes_dict[node_name] == 0:
+                    del self.expected_lifecycle_nodes_dict[node_name]
```

Comment:
Is this necessary? Couldn't you just check for all keys in the dict that their value is 0?

Reply:
Good point, I think I like that better because we can have stricter checks if we want, e.g., we received more messages than expected.

Reply:
Python's `any()` should do you well:
```python
any(v == 0 for _, v in dict.iteritems())
```
Or something along those lines